ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/capella/block/BlockProcessorCapella.java

```diff
@@ -133,12 +167,98 @@ public void processBlsToExecutionChangesNoValidation(
     }
   }
 
+  // process_withdrawals
+  public void processWithdrawals(
+      final MutableBeaconState genericState, final Optional<ExecutionPayload> maybePayload)
+      throws BlockProcessingException {
+    final MutableBeaconStateCapella state = MutableBeaconStateCapella.required(genericState);
+    final List<Withdrawal> expectedWithdrawals = getExpectedWithdrawals(state);
+    final ExecutionPayloadCapella executionPayloadCapella =
+        ExecutionPayloadCapella.required(
+            maybePayload.orElseThrow(
+                () ->
+                    new BlockProcessingException(
+                        "ExecutionPayload was not found during block processing.")));
+    final SszList<Withdrawal> payloadWithdrawals = executionPayloadCapella.getWithdrawals();
+    if (expectedWithdrawals.size() != payloadWithdrawals.size()) {
+      throw new BlockProcessingException(
+          "Expected "
+              + expectedWithdrawals.size()
+              + " withdrawals, but payload contained "
+              + payloadWithdrawals.size());
+    }
+    for (int i = 0; i < expectedWithdrawals.size(); i++) {
+      final Withdrawal withdrawal = payloadWithdrawals.get(i);
+      if (!expectedWithdrawals.get(i).equals(withdrawal)) {
+        throw new BlockProcessingException(
+            "Withdrawal "
+                + withdrawal.getIndex()
+                + " does not match expected withdrawal "
+                + expectedWithdrawals.get(i).getIndex());
+      }
+      beaconStateMutators.decreaseBalance(
+          state, withdrawal.getValidatorIndex().intValue(), withdrawal.getAmount());
+    }
+
+    if (payloadWithdrawals.size() > 0) {
+      final Withdrawal latestWithdrawal = payloadWithdrawals.get(payloadWithdrawals.size() - 1);
+      final int nextWithdrawalValidatorIndex =
+          incrementValidatorIndex(
+              latestWithdrawal.getValidatorIndex().intValue(), genericState.getValidators().size());
+      state.setNextWithdrawalIndex(latestWithdrawal.getIndex().increment());
+      state.setNextWithdrawalValidatorIndex(UInt64.valueOf(nextWithdrawalValidatorIndex));
+    }
+  }
+
+  // get_expected_withdrawals
+  private List<Withdrawal> getExpectedWithdrawals(final BeaconStateCapella preState) {
+    final List<Withdrawal> expectedWithdrawals = new ArrayList<>();
+    final WithdrawalSchema withdrawalSchema = schemaDefinitionsCapella.getWithdrawalSchema();
+    final UInt64 epoch = miscHelpers.computeEpochAtSlot(preState.getSlot());
+    final int validatorCount = preState.getValidators().size();
+    UInt64 withdrawalIndex = preState.getNextWithdrawalIndex();
+    int validatorIndex = preState.getNextWithdrawalValidatorIndex().intValue();
+    for (int i = 0;
+        i < validatorCount
+            && expectedWithdrawals.size() < specConfigCapella.getMaxWithdrawalsPerPayload();
+        i++) {
+      final Validator validator = preState.getValidators().get(validatorIndex);
```

Comment:
We're going to have to come back and optimise this. Getting each validator while searching is performance intensive - we will want to do checks that just depend on the balance first and then only get the actual validator if we really need it.  But I think that's worth logging a follow up ticket as we should do benchmarks etc and this will work fine for testnets.

Reply:
added this comment ref. to #6426